crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -33,11 +34,105 @@ use sp_objects::ObjectsApi;
 use sp_runtime::generic::SignedBlock;
 use sp_runtime::traits::{Block as BlockT, CheckedSub, Header, NumberFor, One, Zero};
 use std::future::Future;
+use std::sync::atomic::{AtomicU16, Ordering};
 use std::sync::Arc;
 use subspace_archiving::archiver::{Archiver, NewArchivedSegment};
 use subspace_core_primitives::crypto::kzg::Kzg;
 use subspace_core_primitives::objects::BlockObjectMapping;
-use subspace_core_primitives::{BlockNumber, RecordedHistorySegment, SegmentHeader};
+use subspace_core_primitives::{BlockNumber, RecordedHistorySegment, SegmentHeader, SegmentIndex};
+
+#[derive(Debug)]
+struct SegmentHeadersStoreInner<AS> {
+    aux_store: Arc<AS>,
+    next_key_index: AtomicU16,
+    /// In-memory cache of segment headers
+    cache: Mutex<Vec<SegmentHeader>>,
+}
+
+/// Persistent storage of segment headers
+#[derive(Debug)]
+pub struct SegmentHeadersStore<AS> {
+    inner: Arc<SegmentHeadersStoreInner<AS>>,
+}
+
+impl<AS> Clone for SegmentHeadersStore<AS> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+}
+
+impl<AS> SegmentHeadersStore<AS>
+where
+    AS: AuxStore,
+{
+    const KEY_PREFIX: &[u8] = b"segment-headers";
+    const INITIAL_CACHE_CAPACITY: usize = 1_000;
+
+    /// Create new instance
+    pub fn new(aux_store: Arc<AS>) -> Result<Self, sp_blockchain::Error> {
+        let mut cache = Vec::with_capacity(Self::INITIAL_CACHE_CAPACITY);
+        let mut next_key_index = 0;
+
+        while let Some(segment_headers) =
```

Comment:
It seems that cache loading time will grow over time. Please, add debug messages before and after cache loading with the current cache size to simplify future debugging of this loading time.